### DISCUSSION TITLE
`api/auth/[...nextauth].js` doesn't work with a custom server

### DISCUSSION THREAD

**Discussion by petercrackthecode** [nextauthjs/next-auth#4377] [Help]
### Environment

System:
- OS: macOS 11.6.2
- CPU: (8) x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
-  Memory: 918.67 MB / 16.00 GB
- Shell: 3.2.57 - /bin/bash

Binaries:
- Node: 14.17.0 - /usr/local/bin/node
- Yarn: 1.22.4 - /usr/local/bin/yarn
- npm: 6.14.13 - /usr/local/bin/npm
- Watchman: 2022.01.03.00 - /usr/local/bin/watchman

Browsers:
- Chrome: 100.0.4896.75
- Edge: 100.0.1185.36
- Firefox: 98.0.1
- Firefox Developer Edition: 100.0
- Safari: 15.3

npmPackages:
- next: 12.1.0 => 12.1.0 
- next-auth: ^4.3.1 => 4.3.1 
- react: 17.0.2 => 17.0.2

### Reproduction URL

https://github.com/petercrackthecode/Test-next-auth

### Describe the issue

I'm using a custom server with express on top of Next JS, and utilizing next-auth for authentication. I use Github as the authenticator. However, the `pages/api/auth/[...nextauth].js` route doesn't work. Once you hit the `Sign In` or `Sign Out` button, the browser console logs out errors as `404: api/auth/signin not found` (or respectively, `404: api/auth/signout not found`).

### How to reproduce

1. Clone the Github repository provided: https://github.com/petercrackthecode/Test-next-auth.
2. `cp .env.example .env.local`
3. Add 4 environment variables to `.env.local`: `GITHUB_ID`, `GITHUB_SECRET`, `NEXTAUTH_SECRET`, `NEXTAUTH_URL`.
4. `yarn install`.
5. `yarn dev`.
6. At `localhost:3000`, click on the `Sign In` button and notice the error shown in the browser console.

### Expected behavior

The user is directed to the Sign In screen with an option to sign in with Github.

**Comment by ThangHuuVu** [comment]
I couldn't open the reproduction URL, could you make it public first? 
On a side note, check https://github.com/nextauthjs/next-auth/issues/531#issuecomment-720261168 to see if it helps.

**Comment by petercrackthecode** [comment]
Hey @ThangHuuVu, thank you for letting me know. I've just made the repo public here: https://github.com/petercrackthecode/Test-next-auth. Would highly appreciate any help!

**Comment by ThangHuuVu** [comment]
From that issue #531, I found a working example: https://github.com/zenflow/hasura-node-monolith-example/blob/2d95df91a8f049c065824cabfbdaacabb16be6b7/auth/routes/next-auth.js
Basically, you don't need to write [...nextauth.jsx] file, but handle NextAuth.js routing in the custom server. Hopefully, this is helpful to you! 🙏

**Comment by petercrackthecode** [comment]
Thank you for the post @ThangHuuVu. Unfortunately, the version mentioned in the example was `^3.29.0`, while I'm using the current v4. I tried to downgrade my next-auth version to the same version as the author's but still had the not found error.

**Comment by cdfornari** [comment]
@petercrackthecode I'm having the same problem in production, did you fix it?

**Reply by petercrackthecode** [reply]
I've made several changes in the repo and it works: https://github.com/petercrackthecode/Test-next-auth/blob/main/server/index.js

The most important changes were made in the `server/index.js` file.

Custom servers don't work with normal serverless deployments like AWS Amplify & Netlify tho. I'm still figuring out how to solve it.

**Reply by drattansingh** [reply]
Did you get this to work both with ...nextauth and posting with request.body?
